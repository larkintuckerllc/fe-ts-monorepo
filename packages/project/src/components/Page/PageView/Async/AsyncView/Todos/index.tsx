import Todos from './Todos';

export default Todos;
