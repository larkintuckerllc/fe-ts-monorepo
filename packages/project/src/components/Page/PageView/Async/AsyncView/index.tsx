import AsyncView from './AsyncView';

export default AsyncView;
