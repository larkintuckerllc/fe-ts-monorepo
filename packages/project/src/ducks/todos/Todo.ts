import { Record } from 'immutable';

const todoDefault = {
  completed: false,
  id: 0,
  title: 'title',
  userID: 0,
};

export interface TodoJS {
  completed: boolean;
  id: number;
  title: string;
  userID: number;
}

export default class Todo extends Record(todoDefault) {
  constructor(params: TodoJS) {
    super(params);
  }
  public get<T extends keyof TodoJS>(value: T): TodoJS[T] {
    return super.get(value);
  }
}
