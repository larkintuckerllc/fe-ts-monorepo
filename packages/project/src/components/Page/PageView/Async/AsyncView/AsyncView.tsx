import Todo from 'DUCKS/todos/Todo';
import { List } from 'immutable';
import React, { Component } from 'react';
import Todos from './Todos';

interface AsyncViewProps {
  error: boolean;
  requested: boolean;
  todos: List<Todo>;
  fetchTodos(): void;
}

export default class AsyncView extends Component<AsyncViewProps> {
  public componentDidMount() {
    const { fetchTodos } = this.props;
    fetchTodos();
  }
  public render() {
    const { error, requested, todos } = this.props;
    if (requested) {
      return <div>Requested</div>;
    }
    if (error) {
      return <div>Error</div>;
    }
    if (todos.size === 0) {
      return <div>No Todos</div>;
    }
    return (
      <div>
        <h2>Async</h2>
        <Todos todos={todos.toJS()} />
      </div>
    );
  }
}
