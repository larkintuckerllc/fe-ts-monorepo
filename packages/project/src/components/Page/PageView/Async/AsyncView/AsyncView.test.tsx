import Todo from 'DUCKS/todos/Todo';
import Enzyme, { shallow } from 'enzyme';
import enzymeAdapterReact16 from 'enzyme-adapter-react-16';
import { List } from 'immutable';
import React from 'react';
import AsyncView from './AsyncView';

Enzyme.configure({ adapter: new enzymeAdapterReact16() });

const todoDefault = {
  completed: false,
  id: 0,
  title: 'title',
  userID: 0,
};
const sampleTodo = new Todo(todoDefault);
const sampleTodos = List([sampleTodo]);
const getDefaultProps = () => ({
  error: false,
  fetchTodos: jest.fn(),
  requested: false,
  todos: sampleTodos,
});

describe('Async component', () => {
  it('shallow renders without crashing', () => {
    const { error, fetchTodos, requested, todos } = getDefaultProps();
    shallow(
      <AsyncView error={error} fetchTodos={fetchTodos} requested={requested} todos={todos} />
    );
  });
  it('renders differently with requested', () => {
    const { error, fetchTodos, todos } = getDefaultProps();
    const wrapper = shallow(
      <AsyncView error={error} fetchTodos={fetchTodos} requested={true} todos={todos} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders differently with not requested and error', () => {
    const { fetchTodos, requested, todos } = getDefaultProps();
    const wrapper = shallow(
      <AsyncView error={true} fetchTodos={fetchTodos} requested={requested} todos={todos} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders differently with not requested not error and 0 todos', () => {
    const { error, fetchTodos, requested } = getDefaultProps();
    const todos = List<Todo>([]);
    const wrapper = shallow(
      <AsyncView error={error} fetchTodos={fetchTodos} requested={requested} todos={todos} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls fetchTodos on mount', () => {
    const { error, fetchTodos, requested, todos } = getDefaultProps();
    shallow(
      <AsyncView error={error} fetchTodos={fetchTodos} requested={requested} todos={todos} />
    );
    expect(fetchTodos.mock.calls.length).toBe(1);
  });
});
