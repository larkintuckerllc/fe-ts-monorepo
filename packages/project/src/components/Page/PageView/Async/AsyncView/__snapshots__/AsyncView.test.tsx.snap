// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async component renders differently with not requested and error 1`] = `
<div>
  Error
</div>
`;

exports[`Async component renders differently with not requested not error and 0 todos 1`] = `
<div>
  No Todos
</div>
`;

exports[`Async component renders differently with requested 1`] = `
<div>
  Requested
</div>
`;
